packages/core/src/assets/assets.ts

```diff
@@ -0,0 +1,391 @@
+    )
+  }
+
+  const {assetName, dimensions, format} = match.groups as {
+    assetName: string
+    dimensions: string
+    format: string
+  }
+  return `https://cdn.sanity.io/images/${resolvedProject}/${resolvedDataset}/${assetName}-${dimensions}.${format}`
+}
+
+/**
+ * Asset document subset returned by default asset queries
+ * @public
+ */
+export interface AssetDocumentBase {
+  _id: string
+  _type: 'sanity.imageAsset' | 'sanity.fileAsset'
+  url?: string
+  path?: string
+  originalFilename?: string
+  size?: number
+  metadata?: unknown
+}
+
+/**
+ * Options for querying assets
+ * @public
+ */
+export interface AssetQueryOptions<
+  TDataset extends string = string,
+  TProjectId extends string = string,
+> extends DatasetHandle<TDataset, TProjectId> {
+  /** Restrict to a specific asset type. Defaults to both */
+  assetType?: 'image' | 'file' | 'all'
+  /** Additional GROQ filter expression appended with && (...) */
+  where?: string
+  /** GROQ params */
+  params?: Record<string, unknown>
+  /** Optional ordering, eg: `_createdAt desc` */
+  order?: string
+  /** Limit number of results */
+  limit?: number
+  /** Optional projection body inside \{...\}. Omit to return full documents */
+  projection?: string
+}
+
+function buildAssetsGroq(
+  options: Omit<AssetQueryOptions, 'projectId' | 'dataset' | 'params'>,
+): string {
+  const typeNames =
+    options.assetType === 'image'
+      ? ['sanity.imageAsset']
+      : options.assetType === 'file'
+        ? ['sanity.fileAsset']
+        : ['sanity.imageAsset', 'sanity.fileAsset']
+
+  const typeFilter = `_type in [${typeNames.map((t) => `"${t}"`).join(', ')}]`
+  const where = options.where ? `${typeFilter} && (${options.where})` : typeFilter
+
+  let groq = `*[${where}]`
+  if (options.order) groq += ` | order(${options.order})`
+  if (typeof options.limit === 'number') groq += `[0...${options.limit}]`
+  if (options.projection) groq += `{${options.projection}}`
+  return groq
+}
+
+const assetsStore = createFetcherStore<[AssetQueryOptions], AssetDocumentBase[]>({
+  name: 'Assets',
+  getKey: (instance, options) => {
+    const projectId = options.projectId ?? instance.config.projectId
+    const dataset = options.dataset ?? instance.config.dataset
+    if (!projectId || !dataset) {
+      throw new Error('A projectId and dataset are required to query assets.')
+    }
+    const groq = buildAssetsGroq(options)
+    const paramsKey = hashString(JSON.stringify(options.params || {}))
+    return `${projectId}.${dataset}:assets:${hashString(groq)}:${paramsKey}`
+  },
+  fetcher: (instance) => (options) => {
+    const projectId = options.projectId ?? instance.config.projectId
+    const dataset = options.dataset ?? instance.config.dataset
+    if (!projectId || !dataset) {
+      throw new Error('A projectId and dataset are required to query assets.')
+    }
+
+    const groq = buildAssetsGroq(options)
+    return getClientState(instance, {
```

Comment:
Is there reason not to use the query store for this? It might be more expected behavior.